### PR TITLE
nixos/systemd-lib: fix restart/reloadTriggers when passing paths

### PR DESCRIPTION
**Pull request by illustris**
## Description of changes

When defining nixos configs like `services.mysql.configFile = ./my.cnf;`, mysql's restart behavior will be inconsistent. For example, if the nixos config is applied from a flake, mysql will restart every time a new config is applied as the store path for the flake source changes each time. This is because `builtins.toString` is applied to the service's restartTriggers:
https://github.com/NixOS/nixpkgs/blob/44d0940ea560dee511026a53f0e2e2cde489b4d4/nixos/lib/systemd-lib.nix#L326

`toString` returns a store path when applied to a derivation, but returns the absolute path to the file when applied to a path. This change will use `"${}"` to explicitly load the config file into the nix store.

```nix
nix-repl> x=./a

nix-repl> x
/tmp/a

nix-repl> toString x
"/tmp/a"

nix-repl> "${x}"
"/nix/store/9hhkqc4m4bvc6knmy20p63a1ypm5j2jq-a"
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by rhendric** [comment]
Should this be fixed in systemd-lib.nix instead? I can't think of any case where someone might intend to have a non-store path in X-Restart-Triggers.

**Comment by illustris** [comment]
The behavior of `toString` and `"${}"` are sufficiently different that changing it in systemd-lib.nix will be a little more complicated.

```
{ X-Restart-Triggers = "${pkgs.writeText "X-Restart-Triggers-${name}" (lib.pipe config.restartTriggers [
  lib.flatten
  (map (x: if lib.isPath x then "${x}" else x))
  toString
])}"; } 
```

This is because toString can [directly stringify many more types](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-toString) compared to [string interpolation](https://nixos.org/manual/nix/stable/language/string-interpolation#interpolated-expression).

If you don't see any issues with this, I'll update the PR to do this for both `X-Restart-Triggers` and `X-Reload-Triggers`.

**Comment by rhendric** [comment]
IMO `concatStringsSep " " config.restartTriggers` would be sufficient. I wouldn't worry about integers or nested lists in their `restartTriggers`.

But I'm not the maintainer for that subsystem or anything; this is just a random opinion. @NixOS/systemd should perhaps weigh in.

**Comment by illustris** [comment]
Ignoring integers and nested lists should be fine for all services in nixpkgs, but it is still a breaking change. Using an absolute non-store path has no use, but I can think of situations where someone might want to use numbers as restart or reload triggers. Let's avoid changing the behavior too much. I'll update the PR and wait for the systemd maintainers to comment.

**Comment by ElvishJerricco** [comment]
So, on the one hand, I think the expected use of `restartTriggers` is to have a changed store path trigger a unit restart, so I think `concatStringsSep " " config.restartTriggers` is sufficient for the expected use case.

However, this code handles *more* cases, and isn't problematically complicated. So I don't see a reason not to do it this way. I think it's good as is.

**Review by ElvishJerricco — DISMISSED** [review]

**Comment by flokli** [comment]
I'm not entirely sure if there's cases where we have paths only available on the deployment target, or which contain secrets, so should not end up in the store.

This will always loop them through the store. I'd like to do a careful search before merging this.

**Comment by illustris** [comment]
before:
someone would have to add `restartTriggers = [ "{./secret-that-shouldnt-be-in-the-nix-store.txt}" ]` for it to end up in the nix store, and for the restart trigger to actually work. `restartTriggers = [ ./secret-that-shouldnt-be-in-the-nix-store.txt ]` would not include the secret in the nix store, but also fail to restart the service as expected, making its inclusion pointless.

after:
`restartTriggers = [ ./secret-that-shouldnt-be-in-the-nix-store.txt ]` would include the secret in the nix store, and restart the service as expected.

Anyone that wants to use secrets as their restart triggers either have broken restart triggers, or are already writing secrets into the nix store.

**Comment by flokli** [comment]
I agree with the argumentation. Please still add a little entry to the release notes, explaining the change, just in case someone was using this (wrongly).

**Comment by illustris** [comment]
> > I'm sorry, I try to avoid accepting excessively pedantic or trivial change suggestions to discourage them. If you insist on these changes, can you please explain:
> > ```
> > 1. in what way the inclusion of "individually" there changes the meaning of that sentence
> > ```
> 
> In a flake context these files were also copied into the store, but due to the way of the _entire_ flake contents being copied, and the store path being content-addressed by all contents, that resulted in overeager restarts.

If as you say you want the release notes to not be "squishy" and "raise more questions", including the word "individually" doesn't convey any of what you explained here.
> > ```
> > 2. how "too often, or too little" is more descriptive than what it was previously
> > ```
> 
> Too often as described in the previous section, too little, as in it being a plan path (outside flake usages) that wasn't interpolated into the store at all.

I and anyone else reading this thread will understand what you meant by "too often, or too little", but someone reading this in the release notes without this context will find your version less clear. 
> > ```
> > 3. why "always" is wrong when the idempotent nature of switching to the same config is assumed to be nix's default behavior
> > ```
> 
> It did not trigger a service reload in case you were not changing any of the flake contents, only if you did other changes, that resulted in a new content-addressed hash of the flake contents

Yes, we both know how switching to an existing config works. I was asking why we explicitly need to call attention to nix's well known default behavior of not restarting services that are already running.
> > ```
> > 4. if/why these changes are significant enough to block the PR
> > ```
> 
> I'd like release notes to be correct. As shown in the previous points, the description was not correct. I also removed the flake-specific parts, as there's also other (non-flake) ways to trigger the same undesired behaviour. Hence the correction. You can also keep the flake mentioning as an example to make it easier to grasp, but the explanation of the underlying issue should be correct.

I too would like the release notes to be correct. Your explanations haven't convinced me that the description is incorrect. Even if you have a contrived example that can trigger this, the examples given here give enough clarity about the issue it solved. If you want the explanation to be "coorect", "too often or too little" is very far from that ideal. That being said, I'm interested to see what other ways you know of triggering this behavior.
> > I don't mean any offense, and it's nothing personal. I appreciate anyone taking the time out of their day to review PRs. This is just a policy I've adopted to address the "death by a thousand nitpicks" problem nixpkgs PRs suffers from.
> 
> And I agree, but these release notes are something read by people in half a year again, the current version is a bit squishy and raises more questions, but especially when it comes to dealing with secrets we should apply an extra bit of care.

Again, "individually" and "too often, or too little" are "squishy and raise more questions". As for this concern about secrets that keeps coming up in this thread, more than half of the options in nixos modules have the _potential_ to commit secrets into the nix store. If you add a secret in a nix option, you are adding that secret to the nix store. You can add `systemd.services.<name>.script = "echo my_gpg_private_key"` just as easily as `restartTriggers = [ ./my_gpg_private_key ]`. If this is a concern, please create an RFC to print a warning for every string and path option in nixpkgs.

tl;dr, I am not convinced that your suggestions improve this PR. If anything, I think it makes things less clear. In the absence of more compelling arguments, I will not be making these changes.

**Comment by ElvishJerricco** [comment]
I'm in agreement with @flokli. I think the suggestions are simple and are improvements. I don't think this PR is in need of any further review. I don't think it matters much, but I *do* take issue with the "always trigger" wording. No, it does not always trigger. You call it "default behavior" when switching to an already active generation does absolutely nothing but that doesn't really make sense to me. That's not "default behavior". That is emergent behavior from the fundamentals of how `switch-to-configuration` works, which does *not* identify whether the generations are the same. "Default behavior" implies it's something that happens all the time unless you configure it otherwise, and that's not how I would describe this concern. It does *not* always trigger, because there are cases where you can `nixos-rebuild` and it will not trigger; whether you're activating the same generation or not is irrelevant to understanding what `switch-to-configuration` will do.

**Comment by illustris** [comment]
Consider `Restart=always` in the context of systemd units. "Always" here doesn't mean the unit keeps restarting every millisecond. Words can have more context-specific meanings, outside of just one of their literal meanings. That being said, I have updated the PR to explicitly mention that this happens when switching to a new configuration.

Your complaints about the use of "default" vs "emergent" is pointless pedantry. You understood the meaning of what was said. Would my using the word "emergent" in this discussion about a bug in the systems module have changed anything? And while we're rolling around in pointless pedantry, I'd argue that it is the "default behaviour", because you can (with a bit of effort) change the behaviour of `switch-to-configuration`. My pointing that out achieves nothing, but I think we're well past trying to achieve anything productive when a relatively simple bugfix has devolved into debating linguistic prescriptivism.

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 1]
```diff
@@ -276,6 +276,9 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - Ada packages (libraries and tools) have been moved into the `gnatPackages` scope. `gnatPackages` uses the default GNAT compiler, `gnat12Packages` and `gnat13Packages` use the respective matching compiler version.
 
+- Paths provided as `restartTriggers` and `reloadTriggers` for systemd units will now be copied into the nix store to make the behavior consistent.
+  Previously, `restartTriggers = [ ./config.txt ]` would either always or never trigger a restart when switching to a new configuration, depending on if the source was in a flake or not.
```
Comment (rhendric) [inline nixos/doc/manual/release-notes/rl-2405.section.md:280]: Can we settle on this? It's more explicit about the bugged behavior than either previous iteration.

```suggestion
  Previously, `restartTriggers = [ ./config.txt ]`, if defined in a flake, would trigger a restart when any part of the flake changed; and if not defined in a flake, would never trigger a restart even if the contents of `config.txt` changed.
```

Comment (illustris) [reply nixos/doc/manual/release-notes/rl-2405.section.md:280]: Yep, looks good.
